### PR TITLE
Make non-minimal OP_IF/NOTIF argument non-standard for P2WSH

### PR DESCRIPTION
**Pull request by jl2012**
The argument for OP_IF/NOTIF should be exactly 0x01 or empty vector in P2WSH. Otherwise, a relay node may replace the argument with arbitrary data. This is now applied as a policy but eventually should become a softfork.

This is not applied to non-segwit scripts because people may have P2SH UTXOs that are not complying with the new rules.

### CODE REVIEW

File: src/script/interpreter.h [hunk 1]
```diff
@@ -94,6 +94,10 @@ enum
     // Making v1-v16 witness program non-standard
     //
     SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM = (1U << 12),
+
+    // Require the argument of OP_IF/NOTIF to be exactly 0x01 or empty vector
```
Comment (sipa) [inline src/script/interpreter.h]: Mention that it only applies to witness scripts.

File: src/test/data/script_tests.json [hunk 2]
```diff
@@ -2125,5 +2125,88 @@
 ["0", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME", "CSV fails if stack top bit 1 << 31 is set and the tx version < 2"],
 ["4294967296", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME",
   "CSV fails if stack top bit 1 << 31 is not set, and tx version < 2"],
-["The End"]
+
+["MINIMALIF tests"],
```
Comment (sipa) [inline src/test/data/script_tests.json]: I don't think we need to include any cases whose behaviour is not changed by minimalif.

Comment (jl2012) [inline src/test/data/script_tests.json]: I'm not sure what should(not) be tested. I want to show that the new rules are not applied to non segwit scripts by accident

File: src/script/interpreter.h (hunk 1 above)
Comment (jl2012) [inline src/script/interpreter.h]: will fix in a squash commit later

**Comment by dcousens** [comment]
concept ACK

**Comment by luke-jr** [comment]
Prefer to just have an OP_CASTTOBOOL added. This breaks conditionals on OP_DEPTH and OP_SIZE for example.

**Comment by jl2012** [comment]
OP_CASTTOBOOL can't solve the problem. CastToBool if OP_IF is exactly the source of malleability. We needs to turn a OP_NOP into OP_ISBOOLVERIFY.

In any case, that requires a softfork which can't be done any time soon. What we need now is a quick policy patch. An alternative is to make `SIZE IF` and `DEPTH IF` excluded from this policy

**Comment by NicolasDorier** [comment]
we did segwit so that we don't have to care about malleability in inputs, but not we care about malleability inside witness inputs? o_O

**Comment by btcdrak** [comment]
Relevant discussion https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2016-August/013014.html

**Comment by NicolasDorier** [comment]
ok read the conversation,  I like the proposition of @jl2012 of making it a relay rule for now and deciding whether to make it a SF later. Concept ACK.

**Comment by luke-jr** [comment]
No objection to a policy rule, just wanted to point that out here so it didn't get lost. :)

(OP_ISBOOLVERIFY could be policy-enforced just as well though...?)

**Comment by instagibbs** [comment]
concept ACK as relay rule only

**Comment by jl2012** [comment]
@luke-jr , a "policy opcode" means we could never use that NOP for other purpose again, even if we abandon the softfork plan.

**Comment by luke-jr** [comment]
@jl2012 Not really. It simply means we'd need to wait a few releases from when it's policy-redefined to a straight-disallowed NOP, before it is safe to reuse. Along those lines, I wonder if it would make sense to redefine other unallocated opcodes as NOPs inside segwit now (although I doubt it will be useful in the long run).

**Comment by jl2012** [comment]
@luke-jr you can't redefine policy opcode because people are likely to have utxos with that.

and it won't make sense to create more NOPs. The use of NOPs is really limited as they can't manipulate the stack at all. For example, my most wanted code, OP_CAT, could not be done with and NOP

**Comment by josephpoon** [comment]
Concept ACK. OK with either policy rule or consensus rule, finalized LN bitcoin scripts will be compatible.

To echo jl2012 's comment about having UTXOs, olicy rules to an eventual SF is complicated because doing something as simple as ```OP_DUP 
OP_IF OP_SHA256 <PUSHDATA> OP_EQUAL```
can invalidate scripts (you're computing on the same data in two places, one of which would be constrained in the SF when activated and the other constrained in the script itself). It is theoretically plausible to construct this kind of script for an actual useful purpose (e.g. if OP_FALSE as input, the other branch would compute on the next item on the stack, albeit it may be possible to construct those scripts more optimally without this problem in the first place, the possibility that it can happen with some kind of script is a concern).

**Review by afk11 — APPROVED** [review]
Code review / tested ACK

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -428,6 +428,12 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
                         if (stack.size() < 1)
                             return set_error(serror, SCRIPT_ERR_UNBALANCED_CONDITIONAL);
                         valtype& vch = stacktop(-1);
+                        if (sigversion == SIGVERSION_WITNESS_V0 && flags & SCRIPT_VERIFY_MINIMALIF) {
```
Comment (instagibbs) [inline src/script/interpreter.cpp]: nit: put some () around the flag check for ease of reading

**Review by instagibbs — APPROVED** [review]
utACK e0b94de0dfbec9b305964a594f2e4785de51655d

File: src/script/interpreter.cpp (hunk 3 above)
Comment (jl2012) [inline src/script/interpreter.cpp]: addressed with c72c5b1

**Comment by btcdrak** [comment]
utACK c72c5b1

**Comment by laanwj** [comment]
Code-review ACK c72c5b1

**Review by sipa — COMMENTED** [review]
utACK c72c5b1

**Pull request merged**